torchao/ops.py

```diff
@@ -43,3 +43,108 @@ def _(_in_feats, _weights, _scales, splitK = 1):
     torch._check(OC == _scales.shape[0], lambda: "Dimensions mismatched")
 
     return _in_feats.new_empty((BS, OC))
+
+
+
+def unpack_tensor_core_tiled_layout(packed_w: Tensor, innerKTiles: int) -> Tensor:
+    """
+    Unpacks weights that were packed with `torch.ops.aten._convert_weight_to_int4pack` to original tensor of shape `N x K`.
+
+    Assumes that the packed weights were generated with `torch.ops.aten._convert_weight_to_int4pack` with `innerKTiles = 2 | 4 | 8`"
+
+    Args:
+        packed_w: torch.tensor: 4D tensor with shape (N / 8) x (K / (innerKTiles * 16)) x 32 x innerKTiles, dtype is torch.int32
+        innerKTiles: int
+
+    Returns:
+        torch.tensor of shape is N x K, dtype is torch.int32
+
+    """
+    return torch.ops.torchao.unpack_tensor_core_tiled_layout.default(
+        packed_w=packed_w, innerKTiles=innerKTiles
+    )
+    
+
+@register_custom_op(f"torchao::unpack_tensor_core_tiled_layout")
+def _(packed_w: Tensor, innerKTiles: int) -> Tensor:
+    torch._check(
+        packed_w.dim() == 4,
+        lambda: f"packed weight should be a 42d tensor, got {packed_w.dim()}D",
+    )
+    torch._check(
+        packed_w.dtype is torch.int32,
+        lambda: f"weight must be INT32, got {packed_w.dtype}",
+    )
+    torch._check(
+        innerKTiles == 2 or innerKTiles == 4 or innerKTiles == 8,
+        lambda: "innerKTiles must be 2, 4, or 8",
+    )
+    torch._check(packed_w.size(2) == 32, lambda: "packed weight must have 32 at dim 2")
+    torch._check(
+        packed_w.size(3) == innerKTiles / 2,
+        lambda: "packed weight must have innerKTiles/2 at dim 3",
+    )
+    N = packed_w.size(0) * 8
+    K = packed_w.size(1) * innerKTiles * 16
+
+    return torch.empty((N, K), dtype=torch.int32, device=packed_w.device)
+
+def dequantize_tensor_core_tiled_layout(packed_w: Tensor, scales_and_zeros: Tensor, group_size: int, innerKTiles: int) -> Tensor:
```

Comment:
do we really need two ops btw, can we just have an unpack op and then call dequantize afterwards to achieve the same thing?

Reply:
Yes - for performance since the second fuses unpacking from tensor-core to logical layout and then dequantizes.